akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -985,6 +955,123 @@ trait FlowOps[+Out, +Mat] {
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' any upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def zip[U](right: Graph[SourceShape[U], _]): Repr[(Out, U), Mat] = zipMat(right)(Keep.left)
+
+  /**
+   * Combine the elements of current flow and given [[Source]] into a stream of tuples.
+   */
+  def zipMat[U, Mat2, Mat3](right: Graph[SourceShape[U], Mat2])(matF: (Mat, Mat2) ⇒ Mat3): Repr[(Out, U), Mat3] =
+    this.viaMat(Flow(right) { implicit b ⇒
+      r ⇒
+        import FlowGraph.Implicits._
+        val zip = b.add(Zip[Out, U]())
+        r ~> zip.in1
+        (zip.in0, zip.out)
+    })(matF)
+
+  /**
+   * Put togather the elements of current flow and given [[Source]]
+   * into a stream of combined elements using a combiner function.
+   *
+   * '''Emits when''' all of the inputs has an element available
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' any upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def zipWith[Out2, Out3](right: Graph[SourceShape[Out2], _])(combine: (Out, Out2) ⇒ Out3): Repr[Out3, Mat] =
+    zipWithMat(right)(combine)(Keep.left)
+
+  /**
+   * Put togather the elements of current flow and given [[Source]]
+   * into a stream of combined elements using a combiner function.
+   */
+  def zipWithMat[Out2, Out3, Mat2, Mat3](right: Graph[SourceShape[Out2], Mat2])(combine: (Out, Out2) ⇒ Out3)(matF: (Mat, Mat2) ⇒ Mat3): Repr[Out3, Mat3] =
+    this.viaMat(Flow(right) { implicit b ⇒
+      r ⇒
+        import FlowGraph.Implicits._
+        val zip = b.add(ZipWith[Out, Out2, Out3](combine))
+        r ~> zip.in1
+        (zip.in0, zip.out)
+    })(matF)
+
+  /**
+   * Merge the given [[Source]] to this [[Flow]], taking elements as they arrive from input streams,
+   * picking randomly when several elements ready.
+   *
+   * '''Emits when''' one of the inputs has an element available
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' all upstreams complete
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def merge[U >: Out](right: Graph[SourceShape[U], _]): Repr[U, Mat] =
+    mergeMat(right)(Keep.left)
+
+  /**
+   * Merge the given [[Source]] to this [[Flow]], taking elements as they arrive from input streams,
+   * picking randomly when several elements ready.
+   */
+  def mergeMat[U >: Out, Mat2, Mat3](right: Graph[SourceShape[U], Mat2])(matF: (Mat, Mat2) ⇒ Mat3): Repr[U, Mat3] =
+    this.viaMat(Flow(right) { implicit b ⇒
+      r ⇒
+        import FlowGraph.Implicits._
+        val merge = b.add(Merge[U](2))
+        r ~> merge.in(1)
+        (merge.in(0), merge.out)
+    })(matF)
+
+  /**
+   * Concatenate the given [[Source]] to this [[Flow]], meaning that once this
+   * Flow’s input is exhausted and all result elements have been generated,
+   * the Source’s elements will be produced.
```

Comment:
I think it should be clear that if `this` onError's then the `right` won't be produced.